pkg/utils/managedresources/managedresources.go

```diff
@@ -432,3 +437,60 @@ func CheckIfManagedResourcesExist(ctx context.Context, c client.Client, class *s
 
 	return false, nil
 }
+
+// GetObjects returns the objects which belong to this managed resource.
+func GetObjects(ctx context.Context, c client.Client, namespace, name string) ([]client.Object, error) {
+	var objects []client.Object
+
+	managedResource := &resourcesv1alpha1.ManagedResource{}
+	if err := c.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, managedResource); err != nil {
+		return nil, fmt.Errorf("could not get ManagedResource %q: %w", client.ObjectKey{Namespace: namespace, Name: name}, err)
+	}
+
+	decoder := serializer.NewCodecFactory(c.Scheme()).UniversalDeserializer()
+	for _, secretRef := range managedResource.Spec.SecretRefs {
+		secret := &corev1.Secret{}
+		if err := c.Get(ctx, client.ObjectKey{Name: secretRef.Name, Namespace: managedResource.Namespace}, secret); err != nil {
+			return nil, fmt.Errorf("could not get secret %q: %w", client.ObjectKey{Name: secretRef.Name, Namespace: managedResource.Namespace}, err)
+		}
+		objectsFromSecret, err := extractObjectsFromSecret(decoder, secret)
+		if err != nil {
+			return nil, fmt.Errorf("could not extract objects from secret %q: %w", client.ObjectKeyFromObject(secret), err)
+		}
+		objects = append(objects, objectsFromSecret...)
+	}
+
+	return objects, nil
+}
+
+func extractObjectsFromSecret(decoder runtime.Decoder, secret *corev1.Secret) ([]client.Object, error) {
+	var objects []client.Object
+
+	for dataKey, dataValue := range secret.Data {
+		var data []byte
+
+		if strings.HasSuffix(dataKey, resourcesv1alpha1.BrotliCompressionSuffix) {
+			reader := brotli.NewReader(bytes.NewReader(dataValue))
+			var err error
+			data, err = io.ReadAll(reader)
+			if err != nil {
+				return nil, fmt.Errorf("could not read brotli compressed data from key %q: %w", dataKey, err)
+			}
+		} else {
+			data = dataValue
+		}
+
+		for _, objSerialized := range strings.Split(string(data), "---\n") {
+			if objSerialized == "" {
+				continue
+			}
+			obj, _, err := decoder.Decode([]byte(objSerialized), nil, nil)
+			if err != nil {
+				return nil, fmt.Errorf("could not decode object: %w", err)
+			}
+			objects = append(objects, obj.(client.Object))
+		}
+	}
+
+	return objects, nil
+}
```

Comment:
Can you adapt https://github.com/gardener/gardener/blob/662a919b986e19a5efe01b28a2fc1d4469d36a29/pkg/utils/test/matchers/managedresource.go#L160-L185 now to use this function (instead of duplicating the logic)?